adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -744,41 +978,244 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+    assert(rddToCoPartitionWith.sorted)
+    val destinationPartitionMap = rddToCoPartitionWith.partitionMap.get
+    //number of partitions we will have after repartition
+    val numPartitions = destinationPartitionMap.length
+
+    val finalPartitionedRDD =
+      if (sorted) {
+        // we're working with sorted data here, so it's important to keep things in
+        // the order. We are going to be moving the data around in a Seq to control
+        // the order.
+        rdd.mapPartitionsWithIndex((idx, iter) => {
+          // we need to derive the ReferenceRegion for each T before we do anything
+          // else.
+          inferCorrectReferenceRegionsForPartition(iter, partitionMap.get(idx)).map(f => {
+            // inside the map over the partition we do a map over the partition map
+            // this implementation is in lieu of for...yield
+            destinationPartitionMap.zipWithIndex.map(partition => {
+              // if there is overlap between the partition map and the current value
+              if (f._1.covers(partition._1._1) || f._1.covers(partition._1._2) ||
+                (f._1.compareTo(partition._1._1) >= 0 && f._1.compareTo(partition._1._2) <= 0)) {
+                // we use Options here so scala doesn't lose typing information
+                // it is important to note that the Seq will contain the same data
+                // in the first value of the tuple for every record. This is the
+                // destination partition index.
+                Some((partition._2, f))
+                // we do this so we don't filter out any data at the end or beginning
+                // we don't want to lose data before the join, and this prevents that
+              } else if ((partition._2 == destinationPartitionMap.length - 1 && f._1.compareTo(partition._1._2) > 0) ||
+                (partition._2 == 0 && f._1.compareTo(partition._1._1) < 0)) {
+                Some((partition._2, f))
+              } else None
+              // next we get rid of the None values from the Seq we just made
+            }).filter(_.isDefined)
+          })
+        }, preservesPartitioning = true)
+          // this part is less intuitive. it's important that things stay ordered
+          // for the sake of preventing another (full) sort. when we arrive here
+          // in the code, we have an RDD[Option[Seq[(Int, (ReferenceRegion, T))]]],
+          // which is quite nasty. what we need is for every Seq to be keyed by the
+          // first ReferenceRegion and destination partition index so the
+          // partitioner can assign it correctly.
+          .mapPartitions(iter => {
+            // since each Seq contains the same partition index for each tuple, we
+            // just pull the first one out of the Seq and drop the rest
+            iter.map(f => ((f.head.get._2._1, f.head.get._1), f.map(g => g.get._2)))
+          })
+          .repartitionAndSortWithinPartitions(
```

Comment:
I think it's ok to take out. The performance difference was about 10%. Spark does something with bundling, but I can't seem to find the documentation on that now.